src/cares_wrap.cc

```diff
@@ -593,10 +597,10 @@ void ChannelWrap::EnsureServers() {
   Setup();
 }
 
-
-class QueryWrap : public AsyncWrap {
+template <void (*F)(ares_channel, const char *, int, int, ares_callback, void *)>
+class BaseWrap : public AsyncWrap {
```

Comment:
I’m wondering if it makes sense to do templating on this level, since it introduces a new parallel class hierarchy … maybe we could store the function as a member of `QueryWrap` instead, and pass it to the `QueryWrap` constructor?

That would also allow us to avoid the `this->…` oddities below…

Reply:
@addaleax not too sure how this works in C++ land. Could you throw some more light? For instance, If I pass in the cares function (say `ares_search`) to the base class (say `BaseWrap`), could the result still be a class (instead of what I assume will be an object)?

I think that was the advantage of templating, so I need not touch wherever `QueryWrap` is being used to create other objects. Maybe some code snippets would help me understand better.

Reply:
@SirR4T 

> For instance, If I pass in the cares function (say `ares_search`) to the base class (say `BaseWrap`), could the result still be a class (instead of what I assume will be an object)?

I don’t really understand this, sorry … What do you mean by “pass to the base class”?

> I think that was the advantage of templating, so I need not touch wherever `QueryWrap` is being used to create other objects.

Doing this also duplicates all generated code and 

Basically, what you have is something like this:

```c++
template <int (*F)()>
class Base {
  /* ... */
};

template <typename BaseWrap>
class SubclassWrap1 : BaseWrap {
  /* ... */
};

/* ... */

template <typename Wrap>
void Query(/* ... */) {
  new Wrap(/*...*/);
}

// And then use:
Query<SubclassWrap1<Base<function1>>>();
Query<SubclassWrap1<Base<function2>>>();
Query<SubclassWrap2<Base<function1>>>();
Query<SubclassWrap2<Base<function2>>>();
/* ... */
```

While this works, it seems very complex in terms of the generated class hiearchy.

What I’m thinking does not modify the existing class hierachy, but rather passes along one more argument:

```c++
class Base {
  Base(int (*fn)()) : search_function_(fn) {}
  /* ... */
 private:
  /* ... */
  int (*search_function_)();
};

class SubclassWrap1 : Base {
  SubclassWrap1(..., int (*fn)()) : Base(..., fn) {}
  // Or, alternatively, we could switch to making the constructor forward 
  // all arguments regardless of what they actually are:
  template <typename... Args>
  SubclassWrap(Args&&... args) : Base(std::forward<Args>(args)...) {}
};

template <typename Wrap>
void Query(/* ... */) {
  new Wrap(..., use_function_1 ? function1 : function2);
}
```

(You could of course use a typedef instead of always spelling out the function type here)